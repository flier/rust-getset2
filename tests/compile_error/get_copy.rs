use getset2::Getter;

#[derive(Getter)]
pub struct Struct {
    #[get(copy)] // #[get(copy)] is not allowed for a field that is not implements `Copy`
    field: Foobar,
}

#[derive(Getter)]
#[get(copy)] // #[get(copy)] is not allowed for a field that is not implements `Copy`
pub struct Struct2 {
    field: Foobar,
}

pub struct Foobar;

fn main() {}
