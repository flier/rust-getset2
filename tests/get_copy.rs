use getset2::Getter;

#[derive(Getter)]
struct Foo<'a> {
    /// `fn copy_field(&self) -> usize`
    #[get(copy, mut)]
    copy_field: usize,

    /// `fn copy_ref_field(&self) -> usize`
    #[get(copy, mut)]
    copy_ref_field: &'a usize,
}

#[test]
fn get_copy() {
    let mut foo = Foo {
        copy_field: 0,
        copy_ref_field: &0,
    };

    *foo.copy_field_mut() = 1;
    assert_eq!(foo.copy_field(), 1);

    *foo.copy_ref_field_mut() = &2;
    assert_eq!(foo.copy_ref_field(), 2);
}
