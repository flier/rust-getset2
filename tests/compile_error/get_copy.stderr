error[E0507]: cannot move out of `self.field` which is behind a shared reference
 --> tests/compile_error/get_copy.rs:5:5
  |
5 | /     #[get(copy)] // #[get(copy)] is not allowed for a field that is not implements `Copy`
6 | |     field: Foobar,
  | |_________^ move occurs because `self.field` has type `Foobar`, which does not implement the `Copy` trait

error[E0507]: cannot move out of `self.field` which is behind a shared reference
  --> tests/compile_error/get_copy.rs:12:5
   |
12 |     field: Foobar,
   |     ^^^^^ move occurs because `self.field` has type `Foobar`, which does not implement the `Copy` trait
